### PR TITLE
BlankNode identifier is only uuid

### PR DESCRIPTION
**Pull request by stain**
This is the one of the more radical suggestions for simplifying BlankNode identifiers. 

*tl;dr: BlankNode `localIdentifier` becomes `identifier` and must be a `java.lang.UUID`. The identifier is now globally unique.*

**Note:** This is a branch I made mainly for discussion - I won't give my own +1 at this stage.

Background: #56 #48 and various `[RDF]` threads on [dev@commons](http://mail-archives.apache.org/mod_mbox/commons-dev/201502.mbox/browser).

## Javadoc

For resulting javadoc, see http://stain.github.io/commons-rdf/blanknode-uuid-only/ 
For comparison, I made the latest master javadoc (including #62) at http://stain.github.io/commons-rdf/master/   
(can this be automated somehow/somewhere? Travis-CI to push the docs somewhere?)

[Javadoc for BlankNode](http://stain.github.io/commons-rdf/blanknode-uuid-only/com/github/commonsrdf/api/BlankNode.html#identifier) now is much simpler to read - although perhaps a bit too thin?  

## Discussion

I am not too happy about loosing control over which bnode ntriples-string is ultimately created - but getting rid of the need for `Optional<Graph> localScope` in BlankNodeImpl is a big win. Equality (and not) of blank nodes is now something that should be easy to achieve across implementations. 

This also solves the problem of  adding/removing nodes between graphs. So semantically this feels much cleaner.

I am slightly concerned about requiring [UUID](http://docs.oracle.com/javase/8/docs/api/java/util/UUID.html)s. They are space-efficient and nice for many reasons - but implementations might have already other mechanisms for tracking their internal identifiers of blank nodes - which might necessitate additional fields, mapping table or a v3/v5 MD5/SHA1 name-based UUID.

But I made the Javadoc vague - you can be equals even if you don't have the same uuid, but if you match by uuid, then you always are the same blank node - "local scope" matching not required.

I'll prepare a second branch with a less restrictive `String identifier()`, which needs more guidance about what is a good identifier string.

I must admit this greatly simplifies many of the other classes - except for the construction cases where previously it could do things like `factory.createBlankNode("org1")` which now must be with anonymous bnode or a UUID-one. Now the `RDFTermFactory` gives no control over the desired ntriplestring - would we still want a way to 'suggest' the ntriplesstring, or is that up to each implementation? I would imagine that such an ntriples string should be decided in a Graph-BlankNode situation rather than directly on a BlankNode.

### CODE REVIEW

File: api/src/main/java/com/github/commonsrdf/api/BlankNode.java [hunk 1]
```diff
@@ -43,77 +57,55 @@
...
+	 * It is not a requirement for this UUID <code>identifier</code> 
+	 * to form part of the {@link #ntriplesString()}. 
+	 *
+	 * @see #equals(Object)
 	 *
-	 * @return An internal, system identifier for the {@link BlankNode}.
+	 * @return An unique identifier for the {@link BlankNode}.
 	 */
-	String internalIdentifier();
+	UUID identifier();
```
Comment (ansell) [inline api/src/main/java/com/github/commonsrdf/api/BlankNode.java]: It may be simpler to return a consistent String representation of the UUID, rather than mandating it to be an actual UUID object.

File: api/src/main/java/com/github/commonsrdf/api/BlankNode.java [hunk 2]
```diff
@@ -13,29 +13,43 @@
...
  * A <a href= "http://www.w3.org/TR/rdf11-concepts/#dfn-blank-node" >RDF-1.1
  * Blank Node</a>, as defined by <a href=
  * "http://www.w3.org/TR/rdf11-concepts/#section-blank-nodes" >RDF-1.1 Concepts
  * and Abstract Syntax</a>, a W3C Recommendation published on 25 February 2014.<br>
  * <p>
- * Note that: Blank nodes are disjoint from IRIs and literals. Otherwise, the
- * set of possible blank nodes is arbitrary. RDF makes no reference to any
- * internal structure of blank nodes.
+ * Note: 
+ * <blockquote cite="http://www.w3.org/TR/rdf11-concepts/#dfn-blank-node">
```
Comment (ansell) [inline api/src/main/java/com/github/commonsrdf/api/BlankNode.java]: Using blockquote cite="http://...." seems to cause issues for me with Java-8 Javadoc.

File: api/src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java [hunk 3]
```diff
@@ -392,9 +363,9 @@ public void invalidLiteralLang() throws Exception {
 
 	@Test(expected = Exception.class)
 	public void invalidTriplePredicate() {
-		BlankNode subject = factory.createBlankNode("b1");
-		BlankNode predicate = factory.createBlankNode("b2");
-		BlankNode object = factory.createBlankNode("b3");
+		BlankNode subject = factory.createBlankNode();
+		BlankNode predicate = factory.createBlankNode();
+		BlankNode object = factory.createBlankNode();
 		factory.createTriple(subject, (IRI) predicate, object);
```
Comment (ansell) [inline api/src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java]: Is this just testing for a ClassCastException? The use of expected=Exception.class is not recommended for that reason that you are not sure where the exception comes from (and using expected=SpecificException.class you are still not sure where the exception comes from)

File: api/src/main/java/com/github/commonsrdf/api/BlankNode.java (hunk 1 above)
Comment (afs) [inline api/src/main/java/com/github/commonsrdf/api/BlankNode.java]: On 06/02/15 01:07, Peter Ansell wrote:
> In api/src/main/java/com/github/commonsrdf/api/BlankNode.java
> <https://github.com/commons-rdf/commons-rdf/pull/63#discussion_r24215817>:
>
>>  	 */
>> -	String internalIdentifier();
>> +	UUID identifier();
>
> It may be simpler to return a consistent String representation of the
> UUID, rather than mandating it to be an actual UUID object.

I agree, and we should not pick one unique identifier mechanism when 
there are alternatives.

1. skolemization as described in RDF 1.1
2. UUID (not always small device friendly)
3. If a system is only for a specific use, "unique" maybe more limited 
range.

Having just `internalIdentifier` keeps that flexibility.

Comment (stain) [inline api/src/main/java/com/github/commonsrdf/api/BlankNode.java]: I got the same feeling as I wrote this code, which is why I made the pull
request #64 as a String variation. Slightly more weasel wordy to prevent
"b2" as identifier, but much more flexible.

I think if this identifier is just an additional thing (e.g we also have
skolemize(), then uuid can be an extra.

    UUID uuid();

Otherwise (it is the sole identifier):

    String identifier ();

I would prefer the second solution (#64)

File: api/src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java (hunk 3 above)
Comment (stain) [inline api/src/test/java/com/github/commonsrdf/api/AbstractRDFTermFactoryTest.java]: I don't know, I only updated the factory/constructor parameters for this
patch, I didn't write any new tests. (Perhaps I should..)

Could you raise this separately against the master branch?
